### DISCUSSION TITLE
How to speed up rendering of html documents

### DISCUSSION THREAD

**Discussion by RKonstantinR** [quarto-dev/quarto-cli#7249] [Q&A] [answered]
### Description

In my case I generate html files based on a quarto document and then take pictures of them using webshot2.

Вut quarto_render function has a large overhead for render and saving the html file.

For example, this code runs in 0.014 sec.

``` R
tictoc::tic()
library(dplyr)

mtcars %>% select(params$col_name)
tictoc::toc()
```
But render an html file already takes about 4 seconds.

```
tic(msg = "Render")
quarto_render("repex_html.qmd", quiet = TRUE)
toc()
Render: 4.064 sec elapsed
```

<details>
<summary>repex_html.qmd</summary>

````` 
---
title: "repex_html"
format: html
params: 
    col_name: "cyl"
---

```{r}

library(dplyr)

mtcars %>% select(params$col_name)

```
`````
</details>

Is there a way to speed up rendering?

**Comment by cderv** [comment, marked as answer]
Quarto runs R as part of its processing but also does a lot of other things, and that takes time. 

I would say a fair comparaison would already be to compare with R run in a new background process. 

````R
> tictoc::tic()
> res <- callr::r_safe(function() {
+     params = list()
+     params$col_name = "cyl"
+     library(dplyr)
+     mtcars %>% select(params$col_name)
+     })
> tictoc::toc()
1.52 sec elapsed
````

Then there will be at least **knitr** and **rmarkdown** layer running, with a Pandoc conversion. This adds more second 

````r
> tictoc::tic()
> res <- callr::r_safe(function() {
+     rmarkdown::render("test2.qmd")
+     })
> tictoc::toc()
3.01 sec elapsed
````

Quarto does then do a lots of processing before and after pandoc renders, which adds some timing
````r
> tictoc::tic()
> quarto::quarto_render("test2.qmd", quiet = TRUE)
> tictoc::toc()
4.88 sec elapsed
````

On my environment, 2 sec more. 

We are tracking performance and always trying to improve. Though, you won't be able to make a quarto rendering going as fast as just running R code in a R console. 

Quarto offers some options to manage execution to freeze or cache outputs : https://quarto.org/docs/projects/code-execution.html
But this will not make a lot of difference if you computation code is already quick. 

Hope it helps

**Comment by RKonstantinR** [comment]
Thanks for reply.

Apparently, in my case, the best solution would be to check data relevance before rendering.
